### PR TITLE
[WIP] Single Key Accounts

### PR DESCRIPTION
**Pull request by muattiyah**
This is a preliminary PR for https://github.com/lbryio/lbry/issues/1179

### CODE REVIEW

File: torba/baseaccount.py [hunk 1]
```diff
@@ -175,3 +175,38 @@ def get_private_key(self, chain, index):
 
     def get_balance(self):
         return self.ledger.db.get_balance_for_account(self)
+
+
+class SingleKeyAccount(BaseAccount):
+    def __init__(self, ledger, seed, encrypted, private_key,
+                 public_key, receiving_gap=0, change_gap=0,
+                 receiving_maximum_use_per_address=0,
+                 change_maximum_use_per_address=0):
```
Comment (muattiyah) [inline torba/baseaccount.py]: What are sane default values for `receiving_gap`, `change_gap`, `receiving_maximum_use_per_address`, `change_maximum_use_per_address`?

File: torba/baseaccount.py [hunk 2]
```diff
@@ -175,3 +175,38 @@ def get_private_key(self, chain, index):
...
+        self.ledger = ledger
+        self.seed = seed
+        self.encrypted = encrypted
+        self.private_key = private_key
+        self.public_key = public_key
+
+        self.receiving, self.change = self.keychains = (
+            # (account, parent_key, chain_number, gap, maximum_use_per_address)
+            KeyChain(self, public_key, 0, 0, 0),
```
Comment (muattiyah) [inline torba/baseaccount.py]: What are sane defaults for `gap`, `maximum_use_per_address`?

File: torba/baseaccount.py [hunk 3]
```diff
@@ -175,3 +175,38 @@ def get_private_key(self, chain, index):
...
+        )
+        # We need this to happen here not in BaseAccount.__init__
+        ledger.add_account(self)
+
+    def get_addresses(self):
+        return list(self.ledger.public_key_to_address(self.public_key))
+
+    def get_unused_addresses(self):
+        return self.ledger.db.get_unused_addresses(self, None)
```
Comment (muattiyah) [inline torba/baseaccount.py]: This isn't going to be very useful in a Single Key account.

**Comment by muattiyah** [comment]
Hi @eukreign. I've left some comments that I would love your input on, this PR is mainly "brainstorming" and we can discard it after settling on a solution if it were far off.

I would also like to know whether we should treat the `Single Key` account with a hierarchy like HD wallets or whether we should discard that (not inherit from `BaseAccount`) and implement a new class with methods that make sense for it?

File: torba/baseaccount.py (hunk 2 above)
Comment (eukreign) [reply torba/baseaccount.py]: I would make a new KeyChain like `SingleKeyChain` that always returns the same key.

File: torba/baseaccount.py (hunk 1 above)
Comment (eukreign) [reply torba/baseaccount.py]: Single Key accounts don't have any of those settings.

File: torba/baseaccount.py (hunk 3 above)
Comment (eukreign) [reply torba/baseaccount.py]: It should always return the same key.

**Comment by eukreign** [comment]
Unfortunately, I've started using `account.receiving` and `account.change` in many other places so these have essentially become public APIs as such the Single Key would have to expose those two attributes but the `SingleKeyChain` should always return the same key.

Also, please keep in mind that torba is undergoing active development so things do and will continue to change drastically. I appreciate this PR but it may be a be premature because things will shift before we can merge it and then it'll likely require a lot of changes.

If you would like to contribute to this project, one thing that would be very helpful is if you can get BitcoinCash integration test working, currently it fails due to improperly set signature bit. We don't have instructions yet on how to run the integration tests but if you're interested in working on the bitcoincash issue please respond here and I'll write up a more detailed ticket for it and how to reproduce the issue.

**Comment by eukreign** [comment]
I've had to accelerate adding this feature because we needed it for lbrynet wallet, this is now done.

**Pull request closed without merging**